File: kotlin/src/main/kotlin/ucb/softarch/currencyconverter/bos/ConversionService.kt
package ucb.softarch.currencyconverter.bos

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import okhttp3.OkHttpClient
import okhttp3.Request
import ucb.softarch.currencyconverter.dtos.ExternalConvertDTO
import ucb.softarch.currencyconverter.dtos.GetConversionRequestDTO
import ucb.softarch.currencyconverter.dtos.GetConversionResponseDTO
import ucb.softarch.currencyconverter.utils.HasLogging
import java.math.BigDecimal

class ConversionService constructor(dto: GetConversionRequestDTO) : HasLogging()
{
    private val currencyFrom = dto.from
    private val currencyTo = dto.to
    private val cashAmount = dto.amount
    private val externalServiceURL = "https://api.currencybeacon.com/v1/convert"
    private val apiKey = ""

    fun validate(): Boolean
    {
        if(cashAmount <= BigDecimal.ZERO)
        {
            logger.error("Validation has failed")
            return false;
        }

        logger.info("Successful validation")
        return true;
    }

    fun convert() : GetConversionResponseDTO
    {
        logger.info("Attempt to connect to external service")
        val externalResponse = getExternalConversion()

        return GetConversionResponseDTO(externalResponse.response.to, externalResponse.response.value)
    }

    private fun buildURL(): String
    {
        return "${externalServiceURL}?api_key=${apiKey}&to=${currencyTo}&from=${currencyFrom}&amount=${cashAmount}"
    }

    private fun getExternalConversion() : ExternalConvertDTO
    {
        val client = OkHttpClient()

        val request = Request.Builder().url(buildURL()).build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful)
            {
                logger.error("External request has failed")
                throw Exception("Unexpected response $response")
            }

            val externalResponse = jacksonObjectMapper().readValue<ExternalConvertDTO>(response.body().string())

            if (externalResponse.response.value == BigDecimal.ZERO)
            {
                logger.warn("Given currencies do not exist")
                throw Exception("Target or source currency doesn't exist")
            }

            logger.info("External request has completed")
            return externalResponse
        }
    }
}

File: dotnet/BOs/ConversionService.cs
namespace currencyconverter.BOs
{
    using currencyconverter.Dtos;
    using currencyconverter.Exceptions;
    using RestSharp;
    using System.Net;

    public class ConversionService
    {
        private readonly string _currencyFrom;
        private readonly string _currencyTo;
        private readonly decimal _cashAmount;
        private readonly string _externalServiceURL;
        private readonly string _apiKey;

        private ILogger _logger;

        public ConversionService(GetConversionRequestDTO dto, ILogger logger)
        {
            _cashAmount = dto.Amount;
            _currencyFrom = dto.From;
            _currencyTo = dto.To;
            _externalServiceURL = "https://api.currencybeacon.com/v1/convert";
            _apiKey = "";

            _logger = logger;
        }

        public bool Validate()
        {
            if (_cashAmount <= 0)
            {
                _logger.LogError("Validation has failed");
                return false;
            }

            _logger.LogInformation("Successfull validation");
            return true;
        }

        public GetConversionResponseDTO Convert()
        {
            _logger.LogInformation("Attempt to connect to external service");
            var externalResponse = GetExternalConversion();

            return new GetConversionResponseDTO(externalResponse.Response.To, externalResponse.Response.Value);
        }

        private string BuildURL()
        {
            return $"{_externalServiceURL}?api_key={_apiKey}&to={_currencyTo}&from={_currencyFrom}&amount={_cashAmount}";
        }

        private ExternalConvertDTO GetExternalConversion()
        {
            var client = new RestClient(BuildURL());

            ExternalConvertDTO? response;

            try
            {
                response = client.Get<ExternalConvertDTO>(new RestRequest());
            }
            catch
            {
                _logger.LogError("External service is unavailable");
                throw new ServiceUnavailableException("Service is temporarily unavailable, please try later");
            }

            if (response == null)
            {
                _logger.LogError("Can't process the external service response");
                throw new InternalServerException("We can't process the request at this time");
            }

            if (response.Response.Value == 0)
            {
                _logger.LogWarning("Given currencies do not exist");
                throw new NotFoundException("Target or source currency doesn't exist");
            }

            _logger.LogInformation("External request has completed");
            return response;
        }
    }
}
